core/src/calleeFunctions/index.ts

```diff
@@ -23,27 +23,104 @@ const allCalleeFunctions: Record<CalleeNames, CalleeFunctions> = {
 export const getCalleeData = async function (
     network: string,
     collateralType: string,
+    marketId: string,
     profitAddress: string
 ): Promise<string> {
     const collateral = getCollateralConfigByType(collateralType);
-    const calleeFuctions = allCalleeFunctions[collateral.exchange.callee];
-    if (!calleeFuctions) {
+    const marketData = collateral.exchanges[marketId];
+    if (!marketData || !marketData.callee || !allCalleeFunctions[marketData.callee]) {
         throw new Error(`Unsupported collateral type "${collateralType}"`);
     }
-    return await calleeFuctions.getCalleeData(network, collateral, profitAddress);
+    return await allCalleeFunctions[marketData.callee].getCalleeData(network, collateral, marketId, profitAddress);
 };
 
-const _getMarketPrice = async function (
+export const getMarketDataById = async function (
+    network: string,
+    collateral: CollateralConfig,
+    marketId: string,
+    amount: BigNumber = new BigNumber('1')
+): Promise<MarketData> {
+    const calleeConfig = collateral.exchanges[marketId];
+    if (!allCalleeFunctions[calleeConfig?.callee]) {
+        throw new Error(`Unsupported callee "${calleeConfig?.callee}" for collateral symbol "${collateral.symbol}"`);
+    }
+    let marketUnitPrice: BigNumber;
+    try {
+        marketUnitPrice = await allCalleeFunctions[calleeConfig.callee].getMarketPrice(
+            network,
+            collateral,
+            marketId,
+            amount
+        );
+    } catch (error: any) {
+        const errorMessage = error?.message;
+        marketUnitPrice = new BigNumber(NaN);
+        return {
+            ...calleeConfig,
+            marketUnitPrice,
+            errorMessage,
+        };
+    }
+    return {
+        ...calleeConfig,
+        marketUnitPrice: marketUnitPrice ? marketUnitPrice : new BigNumber(NaN),
+    };
+};
+
+export const getMarketDataRecords = async function (
     network: string,
     collateralSymbol: string,
     amount: BigNumber = new BigNumber('1')
-): Promise<BigNumber> {
+): Promise<Record<string, MarketData>> {
     const collateral = getCollateralConfigBySymbol(collateralSymbol);
-    const calleeFuctions = allCalleeFunctions[collateral.exchange.callee];
-    if (!calleeFuctions) {
-        throw new Error(`Unsupported collateral symbol "${collateralSymbol}"`);
+    let marketDataRecords = {};
+    for (const marketId in collateral.exchanges) {
+        let marketData: MarketData;
+        try {
+            marketData = await getMarketDataById(network, collateral, marketId, amount);
+        } catch (error: any) {
+            marketData = {
+                errorMessage: error?.message,
+                marketUnitPrice: new BigNumber(NaN),
+                route: [], // dummy value: MarketData requires either a route or tokens
+            };
+        }
```

Comment:
If you remember, we want to show the route on the frontend. In case there is no "marketData", the route will be empty and there will be nothing to display.

If you move try/catch down the `getMarketDataById`, it can return correct `route` or `token` based on the collateral config, not just a dummy empty value.

Please do so in the next frontend PR

Reply:
Good question. I guess linked throw condition shouldn't even be possible (eg via static typing), or should be validated during tests as it's strictly depends on the static config, not on the external conditions.